extensions/amp-gfycat/0.1/amp-gfycat.js

```diff
@@ -48,35 +75,205 @@ class AmpGfycat extends AMP.BaseElement {
   }
 
   /** @override */
-  layoutCallback() {
-    const gfyid = user().assert(
+  buildCallback() {
+    this.videoid_ = this.getVideoId_();
+
+    this.playerReadyPromise_ = new Promise(resolve => {
+      this.playerReadyResolver_ = resolve;
+    });
+
+    installVideoManagerForDoc(this.element);
+    Services.videoManagerForDoc(this.element).register(this);
+  }
+
+  /** @override */
+  createPlaceholderCallback() {
+    const placeholder = this.win.document.createElement('amp-img');
+    dev().assert(this.videoid_);
+    const videoid = this.videoid_ || '';
+
+    placeholder.setAttribute('src',
+        'https://thumbs.gfycat.com/' +
+        encodeURIComponent(videoid) + '-poster.jpg');
+    placeholder.setAttribute('layout', 'fill');
+    placeholder.setAttribute('placeholder', '');
+    placeholder.setAttribute('referrerpolicy', 'origin');
+    this.applyFillContent(placeholder);
+
+    return placeholder;
+  }
+
+  /**
+   * @return {string}
+   * @private
+   */
+  getVideoId_() {
+    return user().assert(
         this.element.getAttribute('data-gfyid'),
         'The data-gfyid attribute is required for <amp-gfycat> %s',
         this.element);
-    const noautoplay = this.element.hasAttribute('noautoplay');
+  }
 
-    const iframe = this.element.ownerDocument.createElement('iframe');
-    iframe.setAttribute('frameborder', '0');
+  /** @return {string} */
+  getVideoIframeSrc_() {
+    if (this.videoIframeSrc_) {
+      return this.videoIframeSrc_;
+    }
+
+    const videoid = this.videoid_ || '';
+    let src = 'https://gfycat.com/ifr/' + encodeURIComponent(videoid);
+
+    const params = getDataParamsFromAttributes(this.element);
 
-    let src = 'https://gfycat.com/ifr/' + encodeURIComponent(gfyid);
+    const noautoplay = this.element.hasAttribute('noautoplay');
     if (noautoplay) {
-      src += '?autoplay=0';
+      params['autoplay'] = '0';
     }
+    src = addParamsToUrl(src, params);
+    return this.videoIframeSrc_ = src;
+  }
 
+  /** @override */
+  layoutCallback() {
+    const iframe = this.element.ownerDocument.createElement('iframe');
+    const src = this.getVideoIframeSrc_();
+
+    iframe.setAttribute('frameborder', '0');
     iframe.src = src;
     this.applyFillContent(iframe);
-    this.element.appendChild(iframe);
     this.iframe_ = iframe;
 
-    return this.loadPromise(iframe);
+    this.unlistenMessage_ = listen(
+        this.win,
+        'message',
+        this.handleGfycatMessages_.bind(this)
+    );
+
+    this.element.appendChild(iframe);
+    const loaded = this.loadPromise(this.iframe_).then(() => {
+      this.element.dispatchCustomEvent(VideoEvents.LOAD);
+    });
+    this.playerReadyResolver_(loaded);
+    return loaded;
```

Comment:
just do:
```
return this.loadPromise(this.iframe_).then(() => {
      this.element.dispatchCustomEvent(VideoEvents.LOAD);
});
```

now you can remove `playerReadyResolver_` and `playerReadyPromsie_`